src/libcore/iter/traits.rs

```diff
@@ -1000,3 +1000,27 @@ pub unsafe trait TrustedLen : Iterator {}
 
 #[unstable(feature = "trusted_len", issue = "37572")]
 unsafe impl<I: TrustedLen + ?Sized> TrustedLen for &mut I {}
+
+/// A specialized trait designed to improve the estimates used when preallocating collections in
+/// cases where `size_hint` is too conservative. For instance, when collecting into an `Option` or a
+/// `Result`, the most common outcome is a non-empty collection, but the protocol allows `size_hint`
+/// to only provide a lower bound of `0`. `OptimisticCollect` can be specialized for such cases in
+/// order to optimize the creation of the resulting collections without breaking `Iterator` rules.
+#[unstable(feature = "optimistic_collect", issue = "00000")]
+pub trait OptimisticCollect: Iterator {
+    /// Provides an estimate of the size of the iterator for the purposes of preallocating
+    /// collections that can be built from it. By default it provides the lower bound of
+    /// `size_hint`.
+    #[inline]
+    fn optimistic_collect_count(&self) -> usize;
+}
+
+#[unstable(feature = "optimistic_collect", issue = "00000")]
+impl<I: Iterator> OptimisticCollect for I {
+    default fn optimistic_collect_count(&self) -> usize { self.size_hint().0 }
+}
+
+#[unstable(feature = "optimistic_collect", issue = "00000")]
+impl<I: Iterator> OptimisticCollect for &mut I {
+    default fn optimistic_collect_count(&self) -> usize { (**self).size_hint().0 }
```

Comment:
Do you need this? Shouldn't `&mut I` `where I: Iterator` *already* be handled by the previous impl and the `Iterator` impl for `&mut I`? Or did you mean to call `(**self).optimistic_collect_count()` here?

Reply:
You're right, I guess this is superfluous.

Reply:
You could make it call `optimistic_collect_count` if you want to. Not sure it can affect this PR, but it could be useful, in general.